### DISCUSSION TITLE
No route to host with self-hosted runner

### DISCUSSION THREAD

**Discussion by martin-g** [community/community#53852] [Actions] [answered]
### Select Topic Area

Question

### Body

Hi,

I face a strange issue with my self hosted runner (actions-runner-linux-arm64-2.303.0.tar.gz).
The runner is installed on a machine that is behind firewall and all outbound HTTP(S) connections should go thru HTTP proxy.
As you can see at https://github.com/martin-g/openeuler-ohpc/actions/runs/4810105457/jobs/8577821961 > **Set up job** the proxy is properly detected!
```
Secret source: Actions
Runner is running behind proxy server 'http://192.168.1.218:3712/' for all HTTP requests.
Runner is running behind proxy server 'http://192.168.1.218:3712/' for all HTTPS requests.
```

But the **Setup** job fails with:
```
wget http://obs.openhpc.community:82/OpenHPC3:/3.0:/Factory/openEuler_22.03/OpenHPC3:3.0:Factory.repo -O /etc/yum.repos.d/ohpc-pre-release.repo
--2023-04-27 15:38:22--  http://obs.openhpc.community:82/OpenHPC3:/3.0:/Factory/openEuler_22.03/OpenHPC3:3.0:Factory.repo
Resolving obs.openhpc.community (obs.openhpc.community)... 44.230.128.55
Connecting to obs.openhpc.community (obs.openhpc.community)|44.230.128.55|:82... failed: No route to host.
```
If I execute the same command (`wget http://obs.openhpc.community:82/OpenHPC3:/3.0:/Factory/openEuler_22.03/OpenHPC3:3.0:Factory.repo`) on the machine itself, using the very same OS user that runs the runner, there is no such problem!

Initially it was failing to resolve the IP address of obs.openhpc.community:
```
wget http://obs.openhpc.community:82/OpenHPC3:/3.0:/Factory/openEuler_22.03/OpenHPC3:3.0:Factory.repo -O /etc/yum.repos.d/ohpc-pre-release.repo
--2023-04-27 15:58:38--  http://obs.openhpc.community:82/OpenHPC3:/3.0:/Factory/openEuler_22.03/OpenHPC3:3.0:Factory.repo
Resolving obs.openhpc.community (obs.openhpc.community)... failed: Name or service not known.
wget: unable to resolve host address ‘obs.openhpc.community’
```
 so I had to add a line in /etc/hosts for it. Again there is no such issue/need when I run wget/curl on the command line when connected via SSH to the machine.

It seems the self-hosted runner uses some kind of restricted network resolver ...

Any ideas how to debug it further ?

Thank you!

**Comment by airtower-luna** [comment]
It's pretty obvious that your `wget` command doesn't use the proxy. You'll have to tell it to, the easiest option is to set the [`http_proxy` and `https_proxy` environment variables](https://manpages.debian.org/bullseye/wget/wget.1.en.html#ENVIRONMENT).

**Reply by martin-g** [reply]
Thank you for your answer!

I see that I was not very clear here!

The http_proxy and https_proxy env variables are exported! 
As I said earlier they are also properly detected by the runner:
```
Runner is running behind proxy server 'http://192.168.1.218:3712/' for all HTTP requests.
Runner is running behind proxy server 'http://192.168.1.218:3712/' for all HTTPS requests.
```

I even print them in the CI job - https://github.com/martin-g/openeuler-ohpc/pull/2/files#diff-d1c7d15699306251199becdfceb20cfa5f36cdf40ea43a839794d36a87608ebfR42

**Reply by airtower-luna** [reply]
The runner process knowing the proxy is not the same as `wget` knowing it. The line in your CI file prints `HTTP_PROXY` and `HTTPS_PROXY`, not `http_proxy` and `https_proxy`. Environment variable names are case-sensitive. :wink:

**Reply by martin-g** [reply]
OK! Let me try!

**Reply by martin-g** [reply]
https://github.com/martin-g/openeuler-ohpc/pull/2/files#diff-d1c7d15699306251199becdfceb20cfa5f36cdf40ea43a839794d36a87608ebfR42-R43

https://github.com/martin-g/openeuler-ohpc/actions/runs/4820814163/jobs/8585825395 prints:
```
echo 'HTTP: http://192.168.1.218:3712/'
HTTP: http://192.168.1.218:3712/
+ echo 'HTTPS: http://192.168.1.218:3712/'
...
wget http://obs.openhpc.community:82/OpenHPC3:/3.0:/Factory/openEuler_22.03/OpenHPC3:3.0:Factory.repo -O /etc/yum.repos.d/ohpc-pre-release.repo
--2023-04-27 22:08:48--  http://obs.openhpc.community:82/OpenHPC3:/3.0:/Factory/openEuler_22.03/OpenHPC3:3.0:Factory.repo
Resolving obs.openhpc.community (obs.openhpc.community)... 44.230.128.55
Connecting to obs.openhpc.community (obs.openhpc.community)|44.230.128.55|:82... failed: No route to host.
```

Still no joy!

**Reply by airtower-luna** [reply, marked as answer]
Oh wait. Where *exactly* is `wget` called there? Possibly in the script called with `sudo`? Sudo removes most environment variables before executing the privileged command (with some variation depending on its config). Try to add `--preserve-env=http_proxy,https_proxy` to the `sudo` command line. With that it should either preserve those variables, or give you a "you're not allowed" message. :slightly_smiling_face:

**Reply by martin-g** [reply]
That's it!
I just added another `wget ...` call before the execution of the script and it was successful!
Thank you, @airtower-luna !

**Comment by martin-g** [comment]
Something that might be related: the machine uses dnsmask running on localhost and delegating to 8.8.8.8 (Google DNS).
There is no systemd-resolved in use (and no /etc/resolv.conf).

As I said: there is no problem executing `wget ...` on the command line (connected via SSH with the same OS user that runs `run.sh`).
